nexus/db-queries/src/db/datastore/migration.rs

```diff
@@ -0,0 +1,58 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods on [`Migration`]s.
+
+use super::DataStore;
+use crate::context::OpContext;
+use crate::db::error::public_error_from_diesel;
+use crate::db::error::ErrorHandler;
+use crate::db::model::Migration;
+use crate::db::schema::migration::dsl;
+use crate::db::update_and_check::UpdateAndCheck;
+use crate::db::update_and_check::UpdateStatus;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use chrono::Utc;
+use diesel::prelude::*;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::UpdateResult;
+use uuid::Uuid;
+
+impl DataStore {
+    /// Insert a database record for a migration.
+    pub async fn migration_insert(
+        &self,
+        opctx: &OpContext,
+        migration: Migration,
+    ) -> CreateResult<Migration> {
+        diesel::insert_into(dsl::migration)
+            .values(migration)
+            .on_conflict(dsl::id)
+            .do_update()
+            .set(dsl::time_created.eq(dsl::time_created))
+            .returning(Migration::as_returning())
+            .get_result_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn migration_mark_deleted(
```

Comment:
Good question! You're absolutely right that currently, migration records don't actually get marked as deleted when the migration completes successfully. This is because...I haven't actually figured out when they can safely be deleted --- since I haven't written the part of the update saga that will consume information from the migration table yet, I'm not totally sure whether migration records can be marked as deleted by the migrate saga as soon as the migration completes, or whether they will need to stick around for a subsequent update saga, and the update saga will have to be responsible for deleting them. I probably should have mentioned that in the PR description.

I didn't want  to add code to the migration saga  to mark migration records as deleted that will then get removed in my update saga branch, so I just didn't make it delete them here. Now that I think through it, though, it occurs to me that this might not be a good idea: if we cut a release that includes this change, but *not* the update saga, we might end up shipping code where we create a bunch of migration records that never get deleted and can't be easily pruned from the DB. I think it's probably better to make a completed migration saga remove the migration record _for now_, and revisit it later when writing the update saga.

Reply:
Actually, looking closer, I think it's `notify_instance_updated` that would be responsible for marking these records as deleted, rather than the `instance-migrate` saga; since that saga just *starts* the migration.

Reply:
Thanks! I think this behavior looks good for now. We may want to change it later, e.g. to support an API that lets an operator migrate an instance and poll to determine when migration has finished, but I don't think we know enough about the necessity or shape of that API to try to design for it just yet. I mention it here just so we can keep it in the backs of our heads.